feat: solana solidity javascript client

vbstreetz opened the pull request:
Very much a work in progress.

Features:
- [x] Upload a BPF program
- [x] Interact with an existing on-chain program
- [x] Call a constructor of a solidity contract
- [x] Call a function of solidity contract, eth abi encoding its arguments and eth abi decode the return values or BPF error return
- [x] Call a function as an RPC call, so off-chain
- [x] Properly decode errors (Retrieve the compute units used, any logs written using print(), and retrieve error)
- [x] Events
- [x] Documentation
- [x] Tests
- [x] Example

seanyoung commented:
Just a note to say that Solang uses a new mechanism for return data: https://github.com/hyperledger-labs/solang/pull/485
- This requires latest solana
- It's possible to retrieve revert errors
- It's possible to do rpc call with `simulateTransaction()`
- Typescript code should be much simpler
- See tests in PR for how this works

Events will be implemented soonish too (PRs are in review)

vbstreetz commented:
Perfect. On it..

On Sat, Sep 18, 2021 at 1:27 PM Sean Young ***@***.***> wrote:

> Just a note to say that Solang uses a new mechanism for return data:
> hyperledger-labs/solang#485
> <https://github.com/hyperledger-labs/solang/pull/485>
>
>    - This requires latest solana
>    - It's possible to retrieve revert errors
>    - It's possible to do rpc call with simulateTransaction()
>    - Typescript code should be much simpler
>    - See tests in PR for how this works
>
> Events will be implemented soonish too (PRs are in review)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/solana-labs/solana-solidity.js/pull/2#issuecomment-922255361>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AON72DOWU2QMB2SAFVG2ZH3UCRSSVANCNFSM5DDQPZ7Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

vbstreetz commented:
Btw, am getting the following error in the latest version:
```
'+solana' is not a recognized feature for this target (ignoring feature)
'+solana' is not a recognized feature for this target (ignoring feature)
'+solana' is not a recognized feature for this target (ignoring feature)
LLVM ERROR: unable to allocate function return #1
Abort trap: 6
```

... compiled with:

```cargo build --release```

seanyoung commented:
> Btw, am getting the following error in the latest version:
> 
> ```
> '+solana' is not a recognized feature for this target (ignoring feature)
> '+solana' is not a recognized feature for this target (ignoring feature)
> '+solana' is not a recognized feature for this target (ignoring feature)
> LLVM ERROR: unable to allocate function return #1
> Abort trap: 6
> ```
> 
> ... compiled with:
> 
> `cargo build --release`

This looks like you're not using the special solana flavoured llvm. You need to build with that llvm in your path. Can either:

 - Download llvm from the last release https://github.com/hyperledger-labs/solang/releases/tag/v0.1.8, extract and add to your path
 - Follow the instructions to build it from source: https://solang.readthedocs.io/en/latest/installing.html#building-llvm-from-source
 - Use the latest docker images https://hub.docker.com/repository/docker/hyperledgerlabs/solang/tags (or extract /usr/bin/solang from it)
 - Use the solang built from the last CI (use the one  from the main branch) https://github.com/hyperledger-labs/solang/actions
 
Let me know if I can help in any way

vbstreetz commented:
Awesome. Now getting this error:
```
Transaction simulation failed: Error processing Instruction 0: Program failed to complete 
    Program 5kaYt1S1Juo2vcCtufmsw21eaThz4sUEqJz75GURyXia invoke [1]
    Program 5kaYt1S1Juo2vcCtufmsw21eaThz4sUEqJz75GURyXia consumed 733 of 200000 compute units
    Program failed to complete: ELF error: Unresolved symbol (sol_set_return_data) at instruction #3171 (ELF file offset 0x6230)
    Program 5kaYt1S1Juo2vcCtufmsw21eaThz4sUEqJz75GURyXia failed: Program failed to complete
```

Will explore it more tomorrow..

seanyoung commented:
@vbstreetz this means your solang is working, however you need to use the latest master branch on solana. You could use the edge docker image:
```
docker run --rm -it -p 8899:8899 -p 8900:8900 solanalabs/solana:edge
```

vbstreetz commented:
that worked. thanks! btw https://github.com/hyperledger-labs/solang/pull/495

seanyoung commented:
@vbstreetz thank you for the PR. 

Events are also supported now. See https://github.com/solana-labs/solana/pull/19764

vbstreetz commented:
awesome. added support for:
- subscribing to events via [callbacks](https://github.com/vbstreetz/solana-solidity.js/blob/master/examples/events/tests/events.spec.ts#L46-L72)
- errors (including parsing [compute units used](https://github.com/vbstreetz/solana-solidity.js/blob/master/examples/errors/tests/errors.spec.ts#L53-L55))

vbstreetz commented:
dry run / simulation as well. format is [`contract.simulate.method(...)`](https://github.com/vbstreetz/solana-solidity.js/blob/master/examples/simulate/tests/simulate.spec.ts#L46)

vbstreetz commented:
will now add some unit tests, mostly on the logs parsing...

vbstreetz commented:
Hey there. Managed to check out the above?

seanyoung commented:
@vbstreetz you have made a lot of progress!

So I've taken a first look. There is a lot of good stuff. I do have some questions:

- deploy can revert. This needs code and test
- I'm not sure the interface for calling functions works. The arguments to
  a function could be an arrray or struct:
  `contract.functions.get_max([1,2,3]);`
  Will the code think that the last argument is the pubkeys/seeds etc object?
- print log message cannot be retrieved:

```
contract hello {
    constructor {
        print("Hello, World!");
    }
}
```
This needs some code and test.

vbstreetz commented:
Gotcha. On it..

seanyoung commented:
The latest solang adds a new value field to the transaction, see https://github.com/hyperledger-labs/solang/pull/502

vbstreetz commented:
Awesome. Fixed issue 2 and also added support for subscribing to `Program log:`s: https://github.com/vbstreetz/solana-solidity.js/blob/master/examples/print/tests/print.spec.ts#L16

vbstreetz commented:
Awesome. Pushed a new version at https://www.npmjs.com/package/solana-solidity. Btw, say more about `deploy can revert. This needs code and test`. I think deployment retry attempt logic should be left to the consumer of the lib or?

seanyoung commented:
> Awesome. Pushed a new version at https://www.npmjs.com/package/solana-solidity. Btw, say more about `deploy can revert. This needs code and test`. I think deployment retry attempt logic should be left to the consumer of the lib or?

Yes, a revert should bubble up to the consumer of the lib.

vbstreetz commented:
Gotcha. On it..

seanyoung commented:
@vbstreetz I've tried to use the library this morning. Some observations:
- The example in `Readme.md` does not compile
- How do I run the tests? There is nothing in `package.json` about tests and when I try to run them I get:

```
$ ./node_modules/.bin/tsc && ./node_modules/.bin/mocha $(find . -name \*.spec.ts)
/home/sean/git/solana-solidity.js/examples/erc20/tests/erc20.spec.ts:1
import expect from 'expect';
^^^^^^

SyntaxError: Cannot use import statement outside a module
```
Please can we have a working `npm test`. I can have solang in my path and solana running, that's not a problem.

In the function call, `pubkeys` should be spit into `accounts` and `writable_accounts`.  `Contract.deploy()` also needs an `accounts` and `writable_accounts`. Both also need a `value` field.

vbstreetz commented:
Btw, pushed an update for catching deployment errors here https://github.com/vbstreetz/solana-solidity.js/commit/259408bba88cc14cd7480c799760624c2b3d966b.

vbstreetz commented:
To run test please use the make targets , for example, to run all tests:
```
make test-all-examples
```

To test a specific example:
```
make test-example o=errors
```

To run unit tests:
```
make test-unit
```

vbstreetz commented:
Added support for:
-  accounts and writableAccounts https://github.com/vbstreetz/solana-solidity.js/commit/c426cd4200b3730ac72792c053acdb1246433cf3
- value https://github.com/vbstreetz/solana-solidity.js/commit/bddddb76e1aadc0cbddd62bce7d7ceaf6d58cbb0

seanyoung commented:
@vbstreetz I'm re-writing the solang tests using this library. I'm not done yet but it looks great so far!

Two issues:
- The `Makefile` is totally broken on Linux
- We will probably publish the package under `@solana/solidity.js` so your package will have to be unpublished, I'm afraid.

vbstreetz commented:
Hey! Awesome. Let me unpublish it. I could also transfer ownership. I'll
also test the Makefile issue on a vps..

On Sun, 3 Oct 2021, 10:59 Sean Young, ***@***.***> wrote:

> @vbstreetz <https://github.com/vbstreetz> I'm re-writing the solang tests
> using this library. I'm not done yet but it looks great so far!
>
> Two issues:
>
>    - The Makefile is totally broken on Linux
>    - We will probably publish the package under @solana/solidity.js so
>    your package will have to be unpublished, I'm afraid.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/solana-labs/solana-solidity.js/pull/2#issuecomment-932884201>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AON72DNUDU7NVHQHFBV6E5DUFAEO5ANCNFSM5DDQPZ7Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

vbstreetz commented:
Ok fixed the Makefile issue + warnings https://github.com/vbstreetz/solana-solidity.js/commit/f363b041bb1c357b93d9d1f1a95bb49071526e0a.

seanyoung commented:
@vbstreetz I have to comment out the first line of the makefile on Linux:
```
$ make
Makefile:1: *** Recursive variable 'o' references itself (eventually).  Stop.
```
With that commented out, the makefile and the tests work fine.

The other issue I have is that when I use the library, I have strict typescript mode enabled. With this enabled I issues with various things. The library should work in strict mode - please could you enable strict mode and fix the issues. (just add `"strict": true,` to `tsconfig.json`. Thank you.

vbstreetz commented:
Gotcha. On it.

Btw two questions:
- Are we able to estimate the `space` needed for a contract account?
- And getting some compile errors on [this](https://gist.github.com/vbstreetz/ea0d9bb6f32cc7e99e716ba428d41e0c).

seanyoung commented:
@vbstreetz thanks for the latest changes, good improvements.

The space needed for a contract is known by the compiler but it's not saved to the abi file. I think we need an extend abi file which includes all the abis, names of the contracts, and how much space they need. This is not implemented yet - still have to decide on a format. So this can be added later.

- The library does not work with `@solana/web3.js` version `1.29.2`. This is needed for the ed25519 program to ed25519 signature verifications, also the library should just work with the latest `@solana/web3.js`.
- errors.spec.ts has typescript errors in it. I think to solve this `TxError` has to be exported (`index.ts` needs `export * from './logs';
`)

vbstreetz commented:
Gotcha. Fixed the TxError issue. Although, I see no errors after upgrading to the [latest `@solana/web3.js`](https://github.com/solana-labs/solana-solidity.js/pull/2/commits/ec2b7a2265bd723990ebe64074bbd6869b95b07d#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R27)..

vbstreetz commented:
> @vbstreetz thanks for the latest changes, good improvements.
> 
> The space needed for a contract is known by the compiler but it's not saved to the abi file. I think we need an extend abi file which includes all the abis, names of the contracts, and how much space they need. This is not implemented yet - still have to decide on a format. So this can be added later.
> 
> * The library does not work with `@solana/web3.js` version `1.29.2`. This is needed for the ed25519 program to ed25519 signature verifications, also the library should just work with the latest `@solana/web3.js`.
> * errors.spec.ts has typescript errors in it. I think to solve this `TxError` has to be exported (`index.ts` needs `export * from './logs'; `)

Mind sharing the error you're experiencing with the latest solana/web3.js?

vbstreetz commented:
Unless it was probably just a type's mismatch issue...

seanyoung commented:
> Mind sharing the error you're experiencing with the latest solana/web3.js?

Sorry that was my bad! No problem there.

However, the event interface is not so great unfortunately.
- Solidity programs often have a quite a few different events. We want to be able to listen to any event a program generates. Also you want the address/account that generated the event (from the program log)
- There is no way to get events from simulate
- There is no way to get  print logs from simulate

Print logs from token.simulate.func() is pretty important.

vbstreetz commented:
Gotcha. I guess we change the api to something like?
```ts
try {
   const {result, logs, computeUnitsUsed, events} = await contract.functions.method(...);
} catch (err) {
  console.log(err.message);
  console.log(err.logs);
  console.log(err.computeUnitsUsed);
}
```

seanyoung commented:
@vbstreetz I can't see how the code matches your comment. `contract.function.method({simulate: true})` still returns the return values, not an object with the results, logs, events, computeUnitsUsed.

Also we still have the problem that the event subscription is only for one event type, not any. 

Lastely it should be possible to set the commitment level in the api.

vbstreetz commented:
Oh, i wanted to confirm if we should change to that 🙂. On it..

On Tue, 5 Oct 2021, 17:05 Sean Young, ***@***.***> wrote:

> @vbstreetz <https://github.com/vbstreetz> I can't see how the code
> matches your comment. contract.function.method({simulate: true}) still
> returns the return values, not an object with the results, logs, events,
> computeUnitsUsed.
>
> Also we still have the problem that the event subscription is only for one
> event type, not any.
>
> Lastely it should be possible to set the commitment level in the api.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/solana-labs/solana-solidity.js/pull/2#issuecomment-934446316>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AON72DLVW5KZRYNA2GV56DTUFMA37ANCNFSM5DDQPZ7Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

seanyoung commented:
ah sorry for the misunderstanding :slightly_smiling_face:

vbstreetz commented:
Hey! I've implemented the above including contracts deployment and API:
```ts
const {program, logs, computeUnitsUsed} = await program.deployContract({
    name: '...',
    abi: '...',
    storageSize: 8192 * 8,
    constructorArgs: [],
  });
```
Also made the `space` param mandatory.

vbstreetz commented:
> There is no test for `createProgramAddress` either; not this works properly.

I actually removed that util func since it wasn't being used. Added some tests for the other utils.

seanyoung commented:
>     * ContractDeployResult is missing events
>       I think you meant ContractCallResult? Anyways, fixed.

Events can be emitted during deploy, so yes I meant `ContractDeployResult`.

>     * contract.functions.method({ simulate: true }) does not return events
>       Fixed.

Thanks, looks good.

>     * if a contract fails before a return value is set (e.g. it runs out of contract space)
>       then the error is overwritten with "return data or log not set"
>       Please share an example on this...

I've created a pr with a test case. https://github.com/vbstreetz/solana-solidity.js/pull/1

Essentially you want any error which occurs *before* return  data is set to be caught and not to be overriden by "no return data set".

>     * the event listener callback function should receive an event type with:
>       Made the listener to only subscribe to the [current program logs](https://github.com/vbstreetz/solana-solidity.js/blob/937c28d67193cdd7d00ebf22e95c0e3148e315ae/src/logs.ts#L190-L192). Please say more about what you meant by `decoded fields` and `topics`. Did you mean filtering similar to ethers.js API https://docs.ethers.io/v5/concepts/events/?

What I means that:

```
export type EventCallback = (...args: any) => void;
```
should become
```
export type EventCallback = (event: ethers.utils.LogDescription) => void;
```

also:

```
  public addEventListener(
    abi: ethers.utils.Interface,
    eventName: string,
    callback: EventCallback
  ): number {
```
should become:
```
  public addEventListener(
    abi: ethers.utils.Interface,
    callback: EventCallback
  ): number {
```

`createProgramAddress()` should be used to create a new address and seed for creating contracts at runtime. I've included a test in https://github.com/vbstreetz/solana-solidity.js/pull/1 which needs fixing up.

Thanks for all the great work! These are all the issues I'm aware of.

jordansexton commented on src/contract.ts:
```diff
@@ -0,0 +1,381 @@
+import { encodeSeeds, numToPaddedHex } from './utils';
+import { Program } from './program';
+
+export type ContractFunction<T = any> = (...args: Array<any>) => Promise<T>;
+
+export type ContractDeployOptions = {
+  name: string;
+  abi: string;
+  space: number;
+  constructorArgs?: any[];
+  accounts?: PublicKey[];
+  writableAccounts?: PublicKey[];
+  seeds?: any[];
+  signers?: Keypair[];
+  caller?: PublicKey | undefined;
+  value?: number;
+  simulate?: boolean;
+};
+
+export type ContractTransactionOptions = {
+  accounts?: PublicKey[];
+  writableAccounts?: PublicKey[];
+  seeds?: any[];
+  signers?: Keypair[];
+  caller?: PublicKey | undefined;
+  value?: number;
+  simulate?: boolean;
+};
```
Could these `any` types be more strict? For function and constructor args especially, it would be helpful to have inputs that can be generically typed and known to the caller.

seanyoung replied:
The seeds need types for sure. If the args/constructor args can be more restricted, that would be great.

vbstreetz replied:
Fixed.

jordansexton commented on src/contract.ts:
```diff
@@ -0,0 +1,381 @@
+
+    let hash = ethers.utils.keccak256(Buffer.from(contractName));
+
+    const data = Buffer.concat([
+      contractStorageAccount.publicKey.toBuffer(),
+      caller.toBuffer(),
+      Buffer.from(numToPaddedHex(value), 'hex'),
+      Buffer.from(hash.substr(2, 8), 'hex'),
+      encodeSeeds(seeds),
+      Buffer.from(input.replace('0x', ''), 'hex'),
+    ]);
```
Can you add a note describing the structure of this data?

jordansexton commented on src/contract.ts:
```diff
@@ -0,0 +1,381 @@
+  TransactionInstruction,
+  SYSVAR_CLOCK_PUBKEY,
+} from '@solana/web3.js';
+import { ethers } from 'ethers';
```
Ethers is kind of sizeable but we're only using a small part of it here. It would be great to import the packages needed to enable tree shaking and avoid duplication.

vbstreetz replied:
gotcha. done.

jordansexton commented on src/contract.ts:
```diff
@@ -0,0 +1,381 @@
+      this.contractStorageAccount.publicKey
+    );
+
+    return accountInfo!.data;
```
If `accountInfo` can be nullish, it seems better to throw a known error.

jordansexton commented on src/contract.ts:
```diff
@@ -0,0 +1,381 @@
+   *
+   * @returns
+   */
+  getStorageKeyPair(): Keypair {
+    return this.contractStorageAccount;
+  }
```
A getter with the same name as the property would produce a more obvious API. Whether you call things accounts or keypairs, it should be consistent.

jordansexton commented on src/program.ts:
```diff
@@ -0,0 +1,286 @@
+   * @param signers      List of signers
+   * @returns            Transaction result
+   */
+  public async makeTx(
```
`sendTransaction` would be more consistent with existing APIs. Can you rely on there only being one instruction? Why not accept a `Transaction` instead?

Overall this method seems a bit long and could probably be broken up into two methods, one for simulation and one for sending, with different options passed to each.

seanyoung replied:
We need to be able to add more than one instruction to a transaction. For ed25519 signature verifications, an ed25519 instruction needs to be included (order doesn't matter).
Besides, multiple instructions per instructions is a great feature that should be exposed :)

jordansexton commented on src/program.ts:
```diff
@@ -0,0 +1,286 @@
+          this.connection,
+          new Transaction().add(instruction),
+          signers,
+          {
+            skipPreflight: false,
+            commitment: 'confirmed',
+            preflightCommitment: undefined,
+          }
```
These `SendOptions` should be options to the function.

jordansexton commented on src/program.ts:
```diff
@@ -0,0 +1,286 @@
+      this.connection,
+      transaction,
+      [this.payerAccount, account],
+      {
+        skipPreflight: false,
+        commitment: 'confirmed',
+        preflightCommitment: undefined,
+      }
```
These should be options too. What makes skipping preflight verification desirable here?

jordansexton commented on src/program.ts:
```diff
@@ -0,0 +1,286 @@
+  ): number {
+    return this.logs.addEventListener(abi, callback);
+  }
+
+  /**
+   * Unsubscribes from the given eventName.
+   *
+   * @param listenerId The log listener id
+   */
+  public async removeEventListener(listenerId: number): Promise<void> {
+    return await this.logs.removeEventListener(listenerId);
+  }
+
+  /**
+   * Deploy a new contract in a loaded Solang program
+   *
+   * @param program
+   * @param options
+   * @returns
+   */
+  public async deployContract(
+    options: ContractDeployOptions
+  ): Promise<ContractDeployResult> {
+    return Contract.deploy(this, options);
+  }
+
+  /**
+   * Load a deployed contract
+   *
+   * @param program
+   * @param abiData
+   * @param contractStorageAccount
+   * @returns                      The contract instance
+   */
+  public async getContract(
+    abiData: string,
+    contractStorageAccount: Keypair
+  ): Promise<Contract> {
+    return Contract.get(this, abiData, contractStorageAccount);
+  }
```
All these methods have one line and just flip some arguments around. It feels like they are adding indirection and API surface area without established cause.

jordansexton commented on src/utils.ts:
```diff
@@ -0,0 +1,31 @@
+import { PublicKey } from '@solana/web3.js';
+
+export function encodeSeeds(seeds: any[]): Buffer {
```
`any` is too loose, you're relying on `seed.seed.length` below.

vbstreetz replied:
Fixed.

jordansexton commented on src/program.ts:
```diff
@@ -0,0 +1,286 @@
+   * @param so              Solang program build bundle
+   * @returns               The program
+   */
+  static async get(
```
Why not call this `load`?

jordansexton commented on src/program.ts:
```diff
@@ -0,0 +1,286 @@
+  private logs: LogsParser;
+
+  /**
+   * Load a new Solang program
```
nit: confusing description for a method called `deploy`

jordansexton commented on src/program.ts:
```diff
@@ -0,0 +1,286 @@
+} from './logs';
+
+export class Program {
+  private logs: LogsParser;
```
I think we should prefer `public` or `protected` over `private` in general unless there's a clear reason to encapsulate. This code provides a base layer, so it should be easy to extend.

jordansexton commented on src/contract.ts:
```diff
@@ -0,0 +1,381 @@
+   * @param fragment
+   * @returns
+   */
+  buildCall(fragment: ethers.utils.Fragment): ContractFunction {
+    return (...args: Array<any>) => {
```
What could be done to infer the types here? If the API for every function is effectively
```ts
(...args: any[]) => Promise<any>
```
this pushes a lot of complexity into the application.

seanyoung replied:
@jordansexton what do you think?

jordansexton replied:
Yeah. The codegen of TypeScript when using hardhat-ethers is **really** nice

vbstreetz replied:
gotcha. looks like typechain still uses ethers underneath https://github.com/dethcrypto/TypeChain/blob/7999f466164b7045a8c57e1fd92c667e24c6a71a/packages/target-ethers-v4-test/test/common.ts#L20

vbstreetz replied:
I think we can postpone this for now..

jordansexton commented on src/contract.ts:
```diff
@@ -0,0 +1,381 @@
+        isSigner: false,
+        isWritable: false,
+      },
+      ...accounts.map((pubkey) => ({
+        pubkey,
+        isSigner: false,
+        isWritable: false,
+      })),
+      ...writableAccounts.map((pubkey) => ({
+        pubkey,
+        isSigner: false,
+        isWritable: true,
+      })),
```
This is something I don't know about programs compiled with Solang. Are the extra accounts always guaranteed to be in this order?

seanyoung replied:
Solang doesn't care about the order of the accounts.

vbstreetz replied:
@seanyoung  It actually does.  For example, for some reason, moving these to the end of the list results in an error https://github.com/vbstreetz/solana-solidity.js/blob/master/src/contract.ts#L256-L260.

jordansexton commented on src/contract.ts:
```diff
@@ -0,0 +1,381 @@
+        try {
+          event = this.abi.parseLog(eventData);
+        } catch (e) {
+          // console.log(e);
```
What can cause a silent error here?

seanyoung replied:
Function cannot fail unless buggy, no need for try/catch

vbstreetz replied:
Fixed.

jordansexton commented on src/logs.ts:
```diff
@@ -0,0 +1,345 @@
+              try {
+                event = abi.parseLog(eventData);
+              } catch (e) {
+                // console.log(e);
```
What can cause a silent error here?

seanyoung replied:
If there is a try catch then the catch should be handled. We shouldn't just swallow errors.

vbstreetz replied:
I guess we just log the error for now or? Since it's totally expected for a discovered emitted event log not to be parseable by any of the registered listeners abis...

seanyoung commented on src/program.ts:
```diff
@@ -0,0 +1,286 @@
+      space
+    );
+
+    let account = Keypair.generate();
```
You're right, the account should be caller-provided. In fact, I think it can just be the public key.

vbstreetz replied:
done. though retained the arg as a keypair since we need to pass the signer here https://github.com/solana-labs/solana-solidity.js/pull/2/files/21087f0919e5dd90e510a0ec02ea474a6bb5964d#diff-34464090febe886c2a40aad4e347f95a9821cff69d4a43a793a376a466b3c885R119

vbstreetz commented:
Hey. Am getting this error in that `create_contract` example https://github.com/hyperledger-labs/solang/blob/09842bccd534775725a1223615aba1811fa765bb/stdlib/solana.c#L135. Am assuming the tx here needs 2 transactions? with the first one creating the pda? please help.

vbstreetz commented:
> > Hey. Am getting this error in that `create_contract` example https://github.com/hyperledger-labs/solang/blob/09842bccd534775725a1223615aba1811fa765bb/stdlib/solana.c#L135. Am assuming the tx here needs 2 transactions? with the first one creating the pda account? please help.
> 
> It needs only one transaction, but it needs both the seed and the program derived address. The program derived address must be the first account in the transaction.
> 
> This test is exactly the same as the one in the solang integration tests. See here how to set it up to run correctly: https://github.com/hyperledger-labs/solang/blob/main/integration/solana/create_contract.spec.ts

Cool. Managed to fix, for some reason, order of the `keys` array matters..

vbstreetz commented on src/program.ts:
```diff
@@ -0,0 +1,286 @@
+    payerAccount: Keypair,
+    so: Buffer
+  ): Promise<Program> {
+    const programAccount = Keypair.generate();
```
Gotcha. So there's actually a different function `get` that accepts the program account but I guess let's remove this one...

vbstreetz commented on src/program.ts:
```diff
@@ -0,0 +1,286 @@
+   * @param programAccount  The program's pubkey
+   */
+  constructor(
+    public connection: Connection,
+    public payerAccount: Keypair,
+    public programAccount: Keypair
```
done

seanyoung commented:
@vbstreetz we would like to build upon your work and improve it, so I'm merging this PR. 

However, and I can't stress this enough, **there are review comments here which still need addressing for the grant to be completed**.

vbstreetz commented:
@seanyoung awesome! please point out any specific reviews I may have missed to attend to. maybe we can open a project board to better track the reviews. btw, @jordansexton will be conducting yet another review soon based on our call yesterday.
